Fixing Tracing Metrics Error - Validate run.system_metrics is not none

kevin-m-kent opened the pull request:
# Description

Previously tracing metrics failed when node system metrics is empty. This checks that it is not none before checking the length. 

# All Promptflow Contribution checklist:
- [ X] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [X ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ X] Title of the pull request is clear and informative.
- [ X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [X] Pull request includes test coverage for the included changes.

kevin-m-kent commented:
I had some trouble running these tests locally, I just couldn't get pytest to run properly - it seemed there were a bunch of dependencies that need to be added to run these tests, couldn't get it to work after fiddling with poetry for a while. Is there a best practice or documentation around running these tests locally? Thank you.

zhengfeiwang commented:
> I had some trouble running these tests locally, I just couldn't get pytest to run properly - it seemed there were a bunch of dependencies that need to be added to run these tests, couldn't get it to work after fiddling with poetry for a while. Is there a best practice or documentation around running these tests locally? Thank you.

hi @kevin-m-kent , thanks for your contribution!

for the doc you are looking for, i think [dev setup](https://github.com/microsoft/promptflow/blob/main/docs/dev/dev_setup.md) and [test](https://github.com/microsoft/promptflow/blob/main/docs/dev/replay-e2e-test.md) are what you are looking for; however, we are doing some package layering and structure refactor, so these docs might not that up to date.

i think you can share your problem/error/screenshot here, or which SHALL BE BETTER, open an issue describing what blocks your development, @crazygao , our expert in fundamental area, who is preparing new docs on tests, can help take a look and better resolve your issue.

zhengfeiwang commented on src/promptflow-core/tests/core/unittests/test_metrics.py:
```diff
@@ -0,0 +1,49 @@
+
+import pytest
+from promptflow.core._serving.monitor.metrics import MetricsRecorder
+from logging import logger
```
this should be the line that breaks the CI for now (which can also break your local test as it's import)

kevin-m-kent commented:
Ok I fixed the import stuff and refactored the test. I can see now that it is generating the warning about failing to record metrics when I have the old code and it doesn't have that warning with the new. But, I have not figured out how to make the test fail based on that. I tried various things with caplog but can't seem to access the right logger to find the message to conduct the assertion. Any suggestions @zhengfeiwang ?

kevin-m-kent commented:
Ok it looks like I figured it out. - the get logger function was defaulting to not propagate. Please take a look when you have a chance @zhengfeiwang . It now fails with the old code and passes with the new.

zhengfeiwang commented:
> Ok it looks like I figured it out. - the get logger function was defaulting to not propagate. Please take a look when you have a chance @zhengfeiwang . It now fails with the old code and passes with the new.

the fail of "check enforcer" might result in that some CIs requires our approval, i just approve that, and let's wait for the result.
